addons/product/models/product_template.py

```diff
@@ -967,24 +1023,24 @@ def _get_possible_combinations(self, parent_combination=None, necessary_values=N
 
         necessary_values = necessary_values or self.env['product.template.attribute.value']
         necessary_attributes = necessary_values.mapped('attribute_id')
-        ptal_stack = [self.valid_product_template_attribute_line_ids.filtered(lambda ptal: ptal.attribute_id not in necessary_attributes)]
-        combination_stack = [necessary_values]
-
-        # keep going while we have attribute lines to test
-        while len(ptal_stack):
-            attribute_lines = ptal_stack.pop()
-            combination = combination_stack.pop()
-
-            if not attribute_lines:
-                # full combination, if it's possible return it, otherwise skip it
-                if self._is_combination_possible(combination, parent_combination):
-                    yield(combination)
-            else:
-                # we have remaining attribute lines to consider
-                for ptav in reversed(attribute_lines[0].product_template_value_ids):
-                    ptal_stack.append(attribute_lines[1:])
-                    combination_stack.append(combination + ptav)
 
+        attributes = self.valid_product_template_attribute_line_ids.filtered(lambda ptal: ptal.attribute_id not in necessary_attributes)
+
+        if not attributes and self._is_combination_possible(necessary_values, parent_combination):
+            yield necessary_values
+
+        # compute all possible values for each attribute
+        attributes_values = [
+            attribute.product_template_value_ids
+            for attribute in attributes
+        ]
+
+        for product in self._cartesian_product(attributes_values):
+            combination = sum(product, necessary_values)
```

Comment:
just use `+`, I assume this line is the equivalent to previously `combination_stack.append(combination + ptav)`, it shouldn't change too much

Reply:
No, it's not equivalent. The code you're referring to added only one value. This sum combines all values to make a combination. The second parameter is that start value of the sum.

Reply:
It doesn't really matter though, recordset can be added with `+` no matter how many elements are in them, both syntax should lead to calling `__add__` in models.py

Reply:
It matters because it's not a record set but a list. It doesn't work with `+`.

Reply:
But the method could/should return a recordset, we shouldn't work with lists of records useless specific case (for example in your method to manipulate index of course, but not needed to have list outside of it).

I think this comes also from trying to plug&play `itertools.product` while we are doing a very specific stuff for our use case?

Reply:
The sum is now a `+`.